### ISSUE TITLE
Valid Behavior: RFC 3339 timestamp precision may be different for different log sources

### ISSUE THREAD

**Issue by schlatterbeck** [rsyslog/rsyslog#5332]
### Expected behavior
Timestamps should be the same for local and remote logs

### Actual behavior
Timestamps of local log lines include the sub-seconds part while remote logs do not. This results in different timestamp formats in the same log file.

Example of local log entry:
```
2024-02-16T22:05:52.315463+01:00 tux [...]
```

Example of remote log entry:
```
2024-02-16T22:06:02+01:00 tux1 [...]
```
The host machine is 'tux' while tux + number are virtual machines on the same host. They use syslog/udp for transferring logs on a local network that is not visible outside the host machine.

### Steps to reproduce the behavior
Use the default RFC 3339 format with local and remote logs on debian linux bookworm

### More Information ###
There is my Debian bug-report on the same issue at
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1064385

This may be related to rsyslogd bug report #2488 "Linux agent timestamp, to match Windows agent timestamp"

### Environment
```
rsyslogd  8.2302.0 (aka 2023.02) compiled with:
        PLATFORM:                               x86_64-pc-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Config file:                            /etc/rsyslog.conf
        PID file:                               /run/rsyslogd.pid
        Number of Bits in RainerScript integers: 64
```

Configuration:
```
module(load="imuxsock") # provides support for local system logging
module(load="imklog")   # provides kernel logging support
module(load="imudp")
input(type="imudp" port="514")
$FileOwner root
$FileGroup adm
$FileCreateMode 0640
$DirCreateMode 0755
$Umask 0022
$WorkDirectory /var/spool/rsyslog
$IncludeConfig /etc/rsyslog.d/*.conf
*.*;auth,authpriv.none          -/var/log/syslog
auth,authpriv.*                 /var/log/auth.log
cron.*                          -/var/log/cron.log
kern.*                          -/var/log/kern.log
mail.*                          -/var/log/mail.log
user.*                          -/var/log/user.log
```

**Comment by davidelang** [comment]
Is this the full config? you could have different templates in use for the two 
outputs.

please log with the template RSYSLOG_DebugFormat so that we can see exactly what 
rsyslog has to work with on the two logs.

David Lang

**Comment by schlatterbeck** [comment]
Yes this is the full config. Except that I've now moved back to the "traditional" log format as the default because it is properly supported by debians logcheck package

Here are two log entries with RSYSLOG_DebugFormat, one produced on the VM "tux1", the other on "tux" (the host machine) both logging to the same logfile.

I'm now using the following two additional log lines in the config without a global default template:
```
daemon.*                        -/var/log/testing.log;RSYSLOG_DebugFormat
daemon.*                        -/var/log/testingx.log
```

I'm testing with `logger -p daemon.info Testing from tux` or `tux1`, respectively.
I'm getting with RSYSLOG_DebugFormat in testing.log:
```
Debug line with all properties:
FROMHOST: 'tux', fromhost-ip: '127.0.0.1', HOSTNAME: 'tux', PRI: 30,
syslogtag 'root:', programname: 'root', APP-NAME: 'root', PROCID: '-', MSGID: '-
',
TIMESTAMP: 'Feb 21 14:42:21', STRUCTURED-DATA: '-',
msg: ' Testing from tux'
escaped msg: ' Testing from tux'
inputname: imuxsock rawmsg: '<30>Feb 21 14:42:21 root: Testing from tux'
$!:
$.:
$/:
```

```
Debug line with all properties:
FROMHOST: 'tux1.runtux.com', fromhost-ip: '10.33.33.1', HOSTNAME: 'tux1', PRI: 3
0,
syslogtag 'root:', programname: 'root', APP-NAME: 'root', PROCID: '-', MSGID: '-
',
TIMESTAMP: 'Feb 21 14:45:20', STRUCTURED-DATA: '-',
msg: ' Testing from tux1'
escaped msg: ' Testing from tux1'
inputname: imudp rawmsg: '<30>Feb 21 14:45:20 tux1 root: Testing from tux1'
$!:
$.:
$/:
```

In testingx.log (with the default template) I'm getting:
```
2024-02-21T14:42:21.905275+01:00 tux root: Testing from tux
```

```
2024-02-21T14:45:20+01:00 tux1 root: Testing from tux1
```

**Comment by rgerhards** [comment]
I guess we need a debug log to tackle this, but I am 99.9999% sure rsyslog does the right thing.

First off, note that RFC5424 does not require the number of digits in the subsecond part of the timestamp - it just prohibits more than 6 digits. The number of digits used reflects the accuracy that the message can capture.

See https://datatracker.ietf.org/doc/html/rfc5424#section-6.2.3.1

Note that RFC3339 also does not require subseconds to be present: https://datatracker.ietf.org/doc/html/rfc3339#section-5.8

Looking at the remote timestamp, it does not provide subsecond resolution at all. So no subseconds are present.

The local timestamp looks the same, but rsyslog uses an optimizaton to query the submission time to the local socket from the OS to obtain more precise time stamps. This is optional, but on by default. If enabled, it ignores the timestamp inside the message and used the high-precision time instead. If you turn that optimization off, I would expect that you get no subseconds in that case either.

The proper fix is to ensure that the sender emits subsecond precision. If that is rsyslog as well, we need to look what happened there (maybe it is using the tradtional template?).

This also makes a lot of sense, as blindly stuffing ".000000" in the subsecond part creates a false sense of precision.

**Comment by davidelang** [comment]
look at the rawmsg field of the debug format output, that shows what rsyslog is 
receiving to parse.

David Lang

**Comment by schlatterbeck** [comment]
On Wed, Feb 21, 2024 at 12:36:16PM -0800, David Lang wrote:
> look at the rawmsg field of the debug format output, that shows what
> rsyslog is receiving to parse.

I don't see any difference between the local and the remote log.
(Except that the local log doesn't have a hostname yet)

So it seems the receiving rsyslog is changing the timestamp?

Thanks
Ralf
-- 
Dr. Ralf Schlatterbeck                  Tel:   +43/2243/26465-16
Open Source Consulting                  www:   www.runtux.com
Reichergasse 131, A-3411 Weidling       email: ***@***.***

**Comment by rgerhards** [comment]
> So it seems the receiving rsyslog is changing the timestamp?

see my explanation above on the imuxsock optimization

**Comment by rgerhards** [comment]
> Where do you see that? I don't see any difference between the two debug log entries. Or was this a reference to the code?

rawmsg from your sample: '<30>Feb 21 14:45:20 tux1 root: Testing from tux1'. This is what rsyslog receives.

See it - no subseconds

> Is it possible to turn that off in the configuration to force all timestamps to the shorter format?

SysSock.IgnoreTimestamp param. doc: https://www.rsyslog.com/doc/configuration/modules/imuxsock.html

You can also define your own forwarding templates with any timestamp you want. See doc for that.

The proper thing to do would be to file a bug against logcheck, not rsyslog. That's because logcheck cannot process perfectly valid RFC5424 message streams, at least that's what you say. It's pretty common to have different resolutions of the timestamps.

In other words: you can work around the logcheck bug by using rsyslog config, but it still is a logcheck bug. So it would be better to once and for all fix the issue at the actual problem root.

**Comment by schlatterbeck** [comment]
For others who come here with the same problem: I've modified my client config to use a (builtin) template for forwarding:
```
auth,authpriv.*                       @syslog:514;RSYSLOG_ForwardFormat
*.*;auth,authpriv.none                @syslog:514;RSYSLOG_ForwardFormat
```

Note the template to use after the semicolon.

This makes both, remote and local log entries use the subseconds part in the timestamp.
